proposals/Layer-1.md

```diff
@@ -0,0 +1,390 @@
+WebAssembly exceptions are designed to allow the use of reference counters to
+perform the the garbage collection in the host VM.
+
+To do this, WebAssembly exceptions are immutable once created, to avoid cyclic
+data structures that can't be garbage collected. It also means that exceptions
+can't be stored into linear memory. The rationale for this is twofold:
+
+* For security. Loads
+  and stores do not guarantee that the data read was of the same type as
+  stored. This allows spoofing of exception references that may allow a
+  WebAssembly module to access data it should not know in the host VM.
+  
+* The host VM does not know the layout of data in linear memory, so it can't
+  find places where exception references are stored.
+
+Hence, while an exception reference is a new first class type, this proposal
+disallows their usage in linear memory.
+
+A WebAssembly exception is created by the `except` instruction. Once an
+exception is created, you can throw it with the `throw` instruction. Thrown
+exceptions are handled as follows:
+
+1. They can be caught by a catch block in an enclosing try block of a function
+   body. The caught exception is pushed onto the stack.
+
+1. Throws not caught within a function body continue up the call stack, popping
+   call frames, until an enclosing try block is found.
+   
+1. If the call stack is exhausted without any enclosing try blocks, the host VM
+   defines how to handle the uncaught exception.
+
+### Exceptions
+
+An `exception` is an internal construct in WebAssembly that is maintained by the
+host. WebAssembly exceptions (as opposed to host exceptions) are defined by a
+new `exception section` of a WebAssembly module. The exception section is a list
+of exception types, from which exceptions can be created.
+
+Each exception type has a `type signature`. The type signature defines the list
+of values associated with the exception.
+
+Within the module, exception types are identified by an index into the
+[exception index space](#exception-index-space). This index is referred to as
+the `exception tag`. The `tagged exception type` is the corresponding
+exception type refered to by the exception tag.
+
+Exception types can be imported and exported by adding the appropriate entries
+to the import and export sections of the module. All imported/exported exception
+types must be named to reconcile exception tags between modules.
+
+Exception tags are used by:
+
+1. The `except` instruction which creates a WebAssembly instance of the
+   corresponding tagged exception type, and pushes a reference to it onto the
+   stack.
+
+2. The `if_except` instruction that queries an exception to see if it is an
+   instance of the corresponding tagged exception class, and if true it pushes
+   the corresponding values of the exception onto the stack.
+   
+### The exception refernce data type
+
+Data types are extended to have a new `except_ref` type. The representation of
```

Comment:
In the current doc, `match` extracts wasm value(s) that was thrown, according to its type signature, from `exn_ref`. Then do you think `match` has to do the two-step extraction, `exn_package` -> `exn_ref` -> wasm values?

Reply:
Yes, `match` (I would probably name it `handle`) would take an exn_package and extract the wasm values if the tag matches. I would actually propose getting rid of exn_ref entirely by merging back except and throw.

Reply:
I'd been envisioning this as that from Wasm's perspective, there is only `exn_ref`. This points to an exception of a certain type with associated values specified by the exception's type. Then, embedders are welcome to wrap this in additional data, making a full `exn_package` that has embedder-specific data such as stack traces. Since `exn_ref` is opaque to Wasm, I don't think it particularly matters what else the embedder does with it, right?

That said, I could see value in separating these things out, in case we wanted to use tagged values as first class data separate from exception handling (which someone will probably try to do anyway).

Reply:
@rossberg - oops, we raced. I think we're in agreement in spirit, just using different terms. In my thinking, you do `handle` on the thing that is caught (an `exn_package`), and it gives you the values associated with the exception if it matches).

Reply:
@eholk, a plain exn value makes no sense as an argument for e.g. rethrow, so you'd still want the distinction. They are really different things.

Reply:
I'm getting confused. What is then the first-class wasm type? `exn_ref` or `exn_package`? Would the composition of `exn_package` object be the same for all embedders?

Reply:
@aheejin, if we did separate exception construction and throwing like in the current state of the proposal both would be first-class. How either is represented is likely different for different embedders.

Reply:
@rossberg, if we combine `throw` and `except`, then we don't need to have separate first class objects for the `exn_ref` and `exn_package`, right?

Reply:
@eholk, yes, there would only be the latter.